File: app/src/main/java/com/annalubawa/shoppinglists/ui/items/ItemsRecyclerAdapter.java
package com.annalubawa.shoppinglists.ui.items;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.annalubawa.shoppinglists.databinding.ItemElementBinding;
import com.annalubawa.shoppinglists.domain.model.Item;

import java.util.List;

public class ItemsRecyclerAdapter extends RecyclerView.Adapter<ItemsRecyclerAdapter.ViewHolder> {

    private List<Item> items;
    private ItemClickListener itemListener;

    public ItemsRecyclerAdapter(List<Item> items, ItemClickListener itemListener) {
        this.items = items;
        this.itemListener = itemListener;
    }

    @NonNull
    @Override
    public ItemsRecyclerAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemElementBinding binding = ItemElementBinding
                .inflate(LayoutInflater.from(parent.getContext()), parent, false);

        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ItemsRecyclerAdapter.ViewHolder holder, int position) {
        holder.binding.setItem(items.get(position));
        holder.binding.getRoot().setOnLongClickListener(v -> {
            itemListener.onLongClick(items.get(position));
            return true;
        });
        holder.binding.checkIconView.setOnClickListener(v ->
                itemListener.onCheckIconClick(items.get(position))
        );
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        ItemElementBinding binding;

        public ViewHolder(@NonNull ItemElementBinding binding) {
            super(binding.getRoot());

            this.binding = binding;
        }
    }

    public interface ItemClickListener {
        void onLongClick(Item item);
        void onCheckIconClick(Item item);
    }
}

File: app/src/main/java/com/annalubawa/shoppinglists/ui/items/ItemsRecyclerAdapter.kt
package com.annalubawa.shoppinglists.ui.items

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.annalubawa.shoppinglists.R
import com.annalubawa.shoppinglists.databinding.ItemElementBinding
import com.annalubawa.shoppinglists.domain.model.Item

class ItemsRecyclerAdapter(
    private val items: List<Item>,
    private val itemListener: ItemClickListener
) : RecyclerView.Adapter<ItemsRecyclerAdapter.ItemsViewHolder>(){

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        ItemsViewHolder(
            DataBindingUtil.inflate<ItemElementBinding>(
                LayoutInflater.from(parent.context),
                R.layout.item_element,
                parent,
                false
            )
        )

    override fun onBindViewHolder(holder: ItemsViewHolder, position: Int) {
        holder.binding.item = items[position]
        holder.binding.root.setOnLongClickListener {
            itemListener.onLongClick(items[position])
            true
        }
        holder.binding.checkIconView.setOnClickListener {
            itemListener.onCheckIconClick(items[position])
        }
    }

    override fun getItemCount(): Int  = items.size

    inner class ItemsViewHolder(val binding: ItemElementBinding)
        : RecyclerView.ViewHolder(binding.root)


    interface ItemClickListener {
        fun onLongClick(item: Item)
        fun onCheckIconClick(item: Item)
    }
}